### PR TITLE
[Merged by Bors] - feat: port Data.Polynomial.Degree.Definitions

### PR DESCRIPTION
**Pull request by Komyyy**
---

- [ ] depends on: #2623
- [ ] depends on: #2624

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#2623
* leanprover-community/mathlib4#2624
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#2623~~
* ~~leanprover-community/mathlib4#2624~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: Mathlib/Data/Polynomial/Degree/Definitions.lean [hunk 1]
```diff
@@ -0,0 +1,1582 @@
...
+universe u v
+
+variable {R : Type u} {S : Type v} {a b c d : R} {n m : ℕ}
+
+section Semiring
+
+variable [Semiring R] {p q r : R[X]}
+
+-- Porting note: In Lean3, these are same `coe`, but not in Lean4.
+private theorem _root_.WithBot.some_eq_nat_cast : WithBot.some n = Nat.cast n := rfl
```
Comment (jcommelin) [inline Mathlib/Data/Polynomial/Degree/Definitions.lean:48]: This already exists as `Nat.cast_withBot` in `Mathlib/Data/Nat/Cast/WithTop.lean`

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**